File: sdk/python/bunqueue/simple/rate_gate.py
"""Client-side rate limiter for Simple Mode (mirror of rate-gate.ts).

Throttles job starts to ``max`` per ``duration`` ms window, optionally per
group (``group_key`` names a field of job.data). A job whose window is full
waits, holding its concurrency slot, until the window frees — the same
observable semantics as the official client's worker limiter.
"""

from __future__ import annotations

import threading
import time
from typing import Any, Dict, List


class RateGate:
    def __init__(self, options: Dict[str, Any]) -> None:
        self._max = int(options["max"])
        self._duration = float(options.get("duration") or 1000)
        self._group_key = options.get("group_key") or options.get("groupKey")
        self._windows: Dict[str, List[float]] = {}
        self._lock = threading.Lock()

    def group_for(self, data: Any) -> str:
        """Resolve the rate-limit group for a job's data."""
        if not self._group_key:
            return ""
        if isinstance(data, dict):
            value = data.get(self._group_key)
            if value is not None:
                return str(value)
        return ""

    def acquire(self, group: str) -> None:
        """Block until the group's window has room, then record the start."""
        while True:
            now = time.monotonic() * 1000
            with self._lock:
                self._prune_locked(now)
                window = [t for t in self._windows.get(group, []) if now - t < self._duration]
                if len(window) < self._max:
                    window.append(now)
                    self._windows[group] = window
                    return
                self._windows[group] = window
                oldest = window[0]
            time.sleep(max((oldest + self._duration - now) / 1000.0, 0.01))

    def _prune_locked(self, now: float) -> None:
        """Evict fully-expired groups (high-cardinality group_key safety)."""
        stale = [
            group
            for group, window in self._windows.items()
            if not window or now - window[-1] >= self._duration
        ]
        for group in stale:
            del self._windows[group]

File: sdk/typescript/src/bunqueue/rate-gate.ts
/**
 * Bunqueue Simple Mode — client-side rate limiter for the worker.
 *
 * The official client passes `rateLimit`/`limiter` to the embedded Worker,
 * which throttles job starts to `max` per `duration` window, optionally per
 * group (`groupKey` names a field of job.data). This SDK reproduces the same
 * semantics as a gate awaited before each job runs: a job whose window is
 * full waits (holding its concurrency slot) until the window frees.
 */

import type { RateLimiterOptions } from './types.js';

const sleep = (ms: number) => new Promise<void>((resolve) => setTimeout(resolve, ms));

export class RateGate {
  private readonly max: number;
  private readonly duration: number;
  private readonly groupKey: string | undefined;
  private readonly windows = new Map<string, number[]>();

  constructor(options: RateLimiterOptions) {
    this.max = options.max;
    this.duration = options.duration;
    this.groupKey = options.groupKey;
  }

  /** Resolve the rate-limit group for a job's data. */
  groupFor(data: unknown): string {
    if (!this.groupKey) return '';
    if (typeof data === 'object' && data !== null) {
      const value = (data as Record<string, unknown>)[this.groupKey];
      if (value !== undefined && value !== null) return String(value);
    }
    return '';
  }

  /** Wait until the group's window has room, then record the start. */
  async acquire(group: string): Promise<void> {
    for (;;) {
      const now = Date.now();
      const window = this.windows.get(group) ?? [];
      const fresh = window.filter((t) => now - t < this.duration);
      if (fresh.length < this.max) {
        fresh.push(now);
        this.windows.set(group, fresh);
        return;
      }
      this.windows.set(group, fresh);
      const oldest = fresh[0];
      await sleep(Math.max(oldest + this.duration - now, 10));
    }
  }

  /**
   * Drop groups whose window is fully expired. Called on each acquire cycle
   * boundary by the owner; without it a high-cardinality groupKey (e.g. one
   * group per user id) grows the map forever.
   */
  prune(): void {
    const now = Date.now();
    for (const [group, window] of this.windows) {
      if (window.length === 0 || now - window[window.length - 1] >= this.duration) {
        this.windows.delete(group);
      }
    }
  }
}
